src/WriteStream.h

```diff
@@ -55,27 +58,55 @@ class WriteStream {
     virtual ~WriteStream() = default;
 
 protected:
-    const std::vector<bool>& symbolMask;
+    const std::vector<char>& kindMask;
     const SymbolTable& symbolTable;
+    const std::vector<int>& recordArityMask;
+    const RecordTable& recordTable;
     const bool isProvenance;
     const bool summary;
     const size_t arity;
 
     virtual void writeNullary() = 0;
     virtual void writeNextTuple(const RamDomain* tuple) = 0;
+
     virtual void writeSize(std::size_t size) {
         assert(false && "attempting to print size of a write operation");
     }
+
     template <typename Tuple>
     void writeNext(const Tuple tuple) {
         writeNextTuple(tuple.data);
     }
+
+    void writeValue(std::ostream& os, size_t col, RamDomain repr) {
```

Comment:
This seems specific to the CSV output, not all WriteStreams

Reply:
I played around with this, and there's some refactoring work needed to get it looking nice. I'll put in a note and leave this to a separate PR.